server/controller/src/main/java/ai/starwhale/mlops/datastore/WalManager.java

```diff
@@ -385,6 +420,13 @@ private PopulationStatus populateOutput() {
                 }
                 entry = this.entriesToWrite.getFirst();
             }
+            if (entry.getId() == FLUSH_ID) {
+                synchronized (this.entriesToWrite) {
+                    this.entriesToWrite.removeFirst();
```

Comment:
It may delete the unexpected entry?

Reply:
FLUSH_ID = -12345, which will never appear in normal entries.

Reply:
I didn't make myself clear, I mean that `getFirst` and `removeFirst` are not in the same synchronized block.

Reply:
entriesToWrite is a queue. Entries are removed from the front and appended to the back. Only populateOutput may remove entries, so it is not a problem.